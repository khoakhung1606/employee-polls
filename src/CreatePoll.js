import React, { useEffect, useState } from "react";
import { addQuestion, setNewQuestions } from "./config/dataActions";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";

function CreatePoll(props) {
  const { newQuestion, setNewQuestions, addQuestion } = props;
  const [optionOne, setOptionOne] = useState("");
  const [optionTwo, setOptionTwo] = useState("");
  const [error, setError] = useState("");
  const navigate = useNavigate();

  useEffect(() => {
    if (newQuestion) {
      setNewQuestions(null);
      navigate("/");
    }
  }, [newQuestion, setNewQuestions, navigate]);

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!optionOne || !optionTwo) {
      setError("Both fields are required!");
    } else {
      setError("");
      addQuestion(optionOne, optionTwo);
    }
  };

  return (
    <div style={{ textAlign: "center", marginTop: "50px" }}>
      <h1>Would You Rather</h1>
      <h2>Create Your Own Poll</h2>
      <form onSubmit={handleSubmit}>
        <label data-testid="first_option_label">
          <h4>First Option</h4>
        </label>
        <div>
          <input
            data-testid="first_option_input"
            type="text"
            placeholder="Option One"
            value={optionOne}
            onChange={(e) => setOptionOne(e.target.value)}
            style={{ width: "800px", marginBottom: "10px" }}
          />
        </div>
        <label data-testid="second_option_label">
          <h4>Second Option</h4>
        </label>
        <div>
          <input
            data-testid="second_option_input"
            type="text"
            placeholder="Option Two"
            value={optionTwo}
            onChange={(e) => setOptionTwo(e.target.value)}
            style={{ width: "800px", marginBottom: "20px" }}
          />
        </div>
        {error && <p style={{ color: "red" }}>{error}</p>}
        <button
          data-testid="btn_submit"
          type="submit"
          disabled={!optionOne || !optionTwo}
          style={{
            backgroundColor: !optionOne || !optionTwo ? "#ccc" : "#007bff",
            color: "#fff",
            padding: "10px 20px",
            border: "none",
            borderRadius: "5px",
            width: "100px",
            marginBottom: "20px",
            cursor: !optionOne || !optionTwo ? "not-allowed" : "pointer",
          }}
        >
          Submit
        </button>
      </form>
    </div>
  );
}

const mapStateToProps = (state) => {
  return {
    newQuestion: state.questionReducer.newQuestion,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setNewQuestions: (question) => {
      dispatch(setNewQuestions(question));
    },
    addQuestion: (optionOneText, optionTwoText) => {
      dispatch(addQuestion(optionOneText, optionTwoText));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CreatePoll);
