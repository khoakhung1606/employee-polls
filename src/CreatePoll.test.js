import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import CreatePoll from "./CreatePoll";
import store from "./config/store";

describe("CreatePoll", () => {
  test("should render the CreatePoll component", async () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreatePoll />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toBeDefined();
    expect(component).toMatchSnapshot();
  });

  test("should call addQuestion when both fields are filled and form is submitted", () => {
    const addQuestionMock = jest.fn();
    const setNewQuestionsMock = jest.fn();

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreatePoll />
        </BrowserRouter>
      </Provider>
    );
    const firstOptionLabel = component.getByTestId("first_option_label");
    const firstOptionInput = component.getByTestId("first_option_input");
    const secondOptionLabel = component.getByTestId("second_option_label");
    const secondOptionInput = component.getByTestId("second_option_input");
    const submit = component.getByTestId("btn_submit");

    expect(firstOptionLabel.textContent).toBe("First Option");
    expect(secondOptionLabel.textContent).toBe("Second Option");
    expect(submit.textContent).toBe("Submit");

    fireEvent.change(firstOptionInput, { target: { value: "option 1" } });
    fireEvent.change(secondOptionInput, { target: { value: "option 2" } });
    expect(firstOptionInput.value).toBe("option 1");
    expect(secondOptionInput.value).toBe("option 2");
    expect(component).toMatchSnapshot();
  });

  test("should have submit button disabled when options are empty", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <CreatePoll />
        </BrowserRouter>
      </Provider>
    );
    const firstOptionInput = component.getByTestId("first_option_input");
    fireEvent.change(firstOptionInput, { target: { value: "option 1" } });

    const submitButton = component.getByTestId("btn_submit");
    expect(submitButton).toBeDisabled();
    expect(component).toMatchSnapshot();
  });
});
