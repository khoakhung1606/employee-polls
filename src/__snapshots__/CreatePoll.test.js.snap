// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePoll should call addQuestion when both fields are filled and form is submitted 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="text-align: center; margin-top: 50px;"
      >
        <h1>
          Would You Rather
        </h1>
        <h2>
          Create Your Own Poll
        </h2>
        <form>
          <label
            data-testid="first_option_label"
          >
            <h4>
              First Option
            </h4>
          </label>
          <div>
            <input
              data-testid="first_option_input"
              placeholder="Option One"
              style="width: 800px; margin-bottom: 10px;"
              type="text"
              value="option 1"
            />
          </div>
          <label
            data-testid="second_option_label"
          >
            <h4>
              Second Option
            </h4>
          </label>
          <div>
            <input
              data-testid="second_option_input"
              placeholder="Option Two"
              style="width: 800px; margin-bottom: 20px;"
              type="text"
              value="option 2"
            />
          </div>
          <button
            data-testid="btn_submit"
            style="background-color: rgb(0, 123, 255); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: pointer;"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="text-align: center; margin-top: 50px;"
    >
      <h1>
        Would You Rather
      </h1>
      <h2>
        Create Your Own Poll
      </h2>
      <form>
        <label
          data-testid="first_option_label"
        >
          <h4>
            First Option
          </h4>
        </label>
        <div>
          <input
            data-testid="first_option_input"
            placeholder="Option One"
            style="width: 800px; margin-bottom: 10px;"
            type="text"
            value="option 1"
          />
        </div>
        <label
          data-testid="second_option_label"
        >
          <h4>
            Second Option
          </h4>
        </label>
        <div>
          <input
            data-testid="second_option_input"
            placeholder="Option Two"
            style="width: 800px; margin-bottom: 20px;"
            type="text"
            value="option 2"
          />
        </div>
        <button
          data-testid="btn_submit"
          style="background-color: rgb(0, 123, 255); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: pointer;"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CreatePoll should have submit button disabled when options are empty 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="text-align: center; margin-top: 50px;"
      >
        <h1>
          Would You Rather
        </h1>
        <h2>
          Create Your Own Poll
        </h2>
        <form>
          <label
            data-testid="first_option_label"
          >
            <h4>
              First Option
            </h4>
          </label>
          <div>
            <input
              data-testid="first_option_input"
              placeholder="Option One"
              style="width: 800px; margin-bottom: 10px;"
              type="text"
              value="option 1"
            />
          </div>
          <label
            data-testid="second_option_label"
          >
            <h4>
              Second Option
            </h4>
          </label>
          <div>
            <input
              data-testid="second_option_input"
              placeholder="Option Two"
              style="width: 800px; margin-bottom: 20px;"
              type="text"
              value=""
            />
          </div>
          <button
            data-testid="btn_submit"
            disabled=""
            style="background-color: rgb(204, 204, 204); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: not-allowed;"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="text-align: center; margin-top: 50px;"
    >
      <h1>
        Would You Rather
      </h1>
      <h2>
        Create Your Own Poll
      </h2>
      <form>
        <label
          data-testid="first_option_label"
        >
          <h4>
            First Option
          </h4>
        </label>
        <div>
          <input
            data-testid="first_option_input"
            placeholder="Option One"
            style="width: 800px; margin-bottom: 10px;"
            type="text"
            value="option 1"
          />
        </div>
        <label
          data-testid="second_option_label"
        >
          <h4>
            Second Option
          </h4>
        </label>
        <div>
          <input
            data-testid="second_option_input"
            placeholder="Option Two"
            style="width: 800px; margin-bottom: 20px;"
            type="text"
            value=""
          />
        </div>
        <button
          data-testid="btn_submit"
          disabled=""
          style="background-color: rgb(204, 204, 204); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: not-allowed;"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CreatePoll should render the CreatePoll component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="text-align: center; margin-top: 50px;"
      >
        <h1>
          Would You Rather
        </h1>
        <h2>
          Create Your Own Poll
        </h2>
        <form>
          <label
            data-testid="first_option_label"
          >
            <h4>
              First Option
            </h4>
          </label>
          <div>
            <input
              data-testid="first_option_input"
              placeholder="Option One"
              style="width: 800px; margin-bottom: 10px;"
              type="text"
              value=""
            />
          </div>
          <label
            data-testid="second_option_label"
          >
            <h4>
              Second Option
            </h4>
          </label>
          <div>
            <input
              data-testid="second_option_input"
              placeholder="Option Two"
              style="width: 800px; margin-bottom: 20px;"
              type="text"
              value=""
            />
          </div>
          <button
            data-testid="btn_submit"
            disabled=""
            style="background-color: rgb(204, 204, 204); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: not-allowed;"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="text-align: center; margin-top: 50px;"
    >
      <h1>
        Would You Rather
      </h1>
      <h2>
        Create Your Own Poll
      </h2>
      <form>
        <label
          data-testid="first_option_label"
        >
          <h4>
            First Option
          </h4>
        </label>
        <div>
          <input
            data-testid="first_option_input"
            placeholder="Option One"
            style="width: 800px; margin-bottom: 10px;"
            type="text"
            value=""
          />
        </div>
        <label
          data-testid="second_option_label"
        >
          <h4>
            Second Option
          </h4>
        </label>
        <div>
          <input
            data-testid="second_option_input"
            placeholder="Option Two"
            style="width: 800px; margin-bottom: 20px;"
            type="text"
            value=""
          />
        </div>
        <button
          data-testid="btn_submit"
          disabled=""
          style="background-color: rgb(204, 204, 204); color: rgb(255, 255, 255); padding: 10px 20px; border-radius: 5px; width: 100px; margin-bottom: 20px; cursor: not-allowed;"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
